docker/resource_docker_service_funcs.go

```diff
@@ -0,0 +1,883 @@
+		return err
+	}
+
+	d.SetId("")
+	return nil
+}
+
+/////////////////
+// Helpers
+/////////////////
+func fetchDockerService(ID string, name string, client *dc.Client) (*swarm.Service, error) {
+	apiServices, err := client.ListServices(dc.ListServicesOptions{})
+
+	if err != nil {
+		return nil, fmt.Errorf("Error fetching service information from Docker: %s", err)
+	}
+
+	for _, apiService := range apiServices {
+		if apiService.ID == ID || apiService.Spec.Name == name {
+			return &apiService, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func deleteService(serviceID string, d *schema.ResourceData, client *dc.Client) error {
+	// == 1: get containerIDs of the running service
+	// because they do not exist after the service is deleted
+	serviceContainerIds := make([]string, 0)
+	if _, ok := d.GetOk("destroy_grace_seconds"); ok {
+		filter := make(map[string][]string)
+		filter["service"] = []string{d.Get("name").(string)}
+		tasks, err := client.ListTasks(dc.ListTasksOptions{
+			Filters: filter,
+		})
+		if err != nil {
+			return err
+		}
+		for _, t := range tasks {
+			task, _ := client.InspectTask(t.ID)
+			log.Printf("[INFO] Found container ['%s'] for destroying: '%s'", task.Status.State, task.Status.ContainerStatus.ContainerID)
+			if strings.TrimSpace(task.Status.ContainerStatus.ContainerID) != "" && task.Status.State != swarm.TaskStateShutdown {
+				serviceContainerIds = append(serviceContainerIds, task.Status.ContainerStatus.ContainerID)
+			}
+		}
+	}
+
+	// == 2: delete the service
+	log.Printf("[INFO] Deleting service: '%s'", d.Id())
+	removeOpts := dc.RemoveServiceOptions{
+		ID: serviceID,
+	}
+
+	if err := client.RemoveService(removeOpts); err != nil {
+		return fmt.Errorf("Error deleting service %s: %s", serviceID, err)
+	}
+
+	// == 3: destroy each container after a grace period
+	if v, ok := d.GetOk("destroy_grace_seconds"); ok {
+		for _, containerID := range serviceContainerIds {
+			timeout := v.(int)
+			destroyGraceSeconds := time.Duration(timeout) * time.Second
+			log.Printf("[INFO] Waiting for container: '%s' to exit: max %v", containerID, destroyGraceSeconds)
+			ctx, cancel := context.WithTimeout(context.Background(), destroyGraceSeconds)
+			defer cancel()
+			exitCode, _ := client.WaitContainerWithContext(containerID, ctx)
+			log.Printf("[INFO] Container exited with code [%v]: '%s'", exitCode, containerID)
+
+			removeOpts := dc.RemoveContainerOptions{
+				ID:            containerID,
+				RemoveVolumes: true,
+				Force:         true,
+			}
+
+			log.Printf("[INFO] Removing container: '%s'", containerID)
+			if err := client.RemoveContainer(removeOpts); err != nil {
+				if !strings.Contains(err.Error(), "No such container") {
+					return fmt.Errorf("Error deleting container %s: %s", containerID, err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+//////// Convergers
+func waitOnService(ctx context.Context, client *dc.Client, plainConvergeConfig *convergeConfig, serviceID string) error {
```

Comment:
This function `waitOnService` needs to be documented if it's to be used. There's a lot going on and it seems to provide important functionality. 

From what I see, it seems to provide similar functionality to Terraform's builtin `StateChangeConf.WaitForState()`, part of the `resource` package:

- https://github.com/hashicorp/terraform/blob/master/helper/resource/state.go#L20-L51

It is preferable to refactor any usage of this `waitOnService` to use `WaitForState` instead.

Reply:
Example usage of `WaitForState`:

- https://github.com/terraform-providers/terraform-provider-aws/blob/a18b8052ad83ad6711522e3dc9f6b8efbd298b46/aws/resource_aws_rds_cluster.go#L353-L366

Reply:
Ok I'll take a look at it. Happy that I have tests :)